src/ameba/formatter/sarif_formatter.cr

```diff
@@ -0,0 +1,305 @@
+    def finished(sources) : Nil
+      sarif_rules = Rule.rules.map do |rule|
+        rule_inst = rule.new
+
+        AsSARIF::ReportingDescriptor.new(
+          id: rule_inst.name,
+          name: rule_inst.name,
+          short_description: rule_inst.description,
+          full_description: rule.parsed_doc || "",
+          help_uri: "https://crystal-ameba.github.io/#{rule_inst.name}",
+          default_configuration: AsSARIF::ReportingConfiguration.new(
+            enabled: rule_inst.enabled?,
+            level: AsSARIF::Level.from_severity(rule_inst.severity),
+            parameters: ""
+          ),
+        )
+      end
+
+      sarif_tool = AsSARIF::Tool.new(
+        driver: AsSARIF::ToolComponent.new(
+          name: "ameba",
+          version: Ameba::VERSION,
+          rules: sarif_rules,
+          # TODO(margret): Better version-specific link
+          information_uri: "https://crystal-ameba.github.io/",
+        )
+      )
+
+      sarif_run = AsSARIF::Run.new(
+        tool: sarif_tool
+      )
+
+      sources.each do |source|
+        source.issues.each do |issue|
+          next if issue.disabled?
+
+          start_location = issue.location || issue.end_location
+          end_location = issue.end_location || issue.location
+
+          if start_location && end_location
+            context_snippet =
+              source.lines[(start_location.line_number - 1)...end_location.line_number].join('\n')
+
+            context_region = AsSARIF::ContextRegion.new(
+              start_line: start_location.line_number,
+              start_column: start_location.column_number,
+              end_line: end_location.line_number,
+              end_column: end_location.line_number,
+              snippet: context_snippet,
+              source_language: source.ecr? ? "ECR" : "Crystal",
+            )
+          end
+
+          sarif_run.results << AsSARIF::RunResult.new(
+            message: issue.message,
+            rule_id: issue.rule.name,
+            rule_index: sarif_rules.index! { |rule| rule.name == issue.rule.name },
+            level: AsSARIF::Level.from_severity(issue.rule.severity),
+            locations: [AsSARIF::Location.new(
+              uri: "file://#{source.fullpath}",
+              context_region: context_region,
+              start_location: start_location,
+              end_location: end_location
+            )]
+          )
+        end
+      end
+
+      sarif_result = AsSARIF::Result.new
+      sarif_result.runs.push(sarif_run)
+
+      sarif_result.to_json(@output)
+    end
+  end
+
+  private module AsSARIF
+    struct Result
+      property schema : String = "https://www.schemastore.org/sarif-2.1.0.json"
+      property version : String = "2.1.0"
+      property runs = Array(Run).new
+
+      def initialize
+      end
+
+      def to_json(json)
+        {
+          "$schema": schema,
+          version:   version,
+          runs:      runs,
+        }.to_json(json)
+      end
+    end
```

Comment:
Using `JSON::Serializable` should be preferred for the simple record definitions.

```suggestion
    struct Result
      include JSON::Serializable
      
      @[JSON::Field(key: "$schema")]
      property schema = "https://www.schemastore.org/sarif-2.1.0.json"
      property version = "2.1.0"
      property runs = [] of Run
    end
```

ditto elsewhere